File: frontend/src/lib/dmxLiveWidget.ts
import type {DMXChannel, DMXChannelType, JSONMap} from "../types/controller";
import {parseFixtureEntries} from "./dmxLiveMap";

export type DMXLiveWidget =
    | "auto"
    | "hidden"
    | "slider"
    | "slotSlider"
    | "buttons"
    | "buttonSlider"
    | "colorWheel"
    | "goboWheel"
    | "shutterModes";

export type LiveSlotKind = "button" | "slider";

export const DMX_LIVE_WIDGET_OPTIONS: { value: DMXLiveWidget; label: string }[] = [
    {value: "auto", label: "Auto"},
    {value: "hidden", label: "Hidden"},
    {value: "slider", label: "Slider"},
    {value: "slotSlider", label: "Slot slider"},
    {value: "buttons", label: "Buttons"},
    {value: "buttonSlider", label: "Switch + slider"},
    {value: "colorWheel", label: "Color wheel"},
    {value: "goboWheel", label: "Gobo wheel"},
    {value: "shutterModes", label: "Shutter modes"},
];

const FINE_TYPES = new Set<DMXChannelType>(["panFine", "tiltFine", "dimmerFine", "zoomFine", "focusFine", "irisFine", "frostFine"]);

const LIVE_MAPPABLE_TYPES = new Set<DMXChannelType>([
    "pan",
    "panFine",
    "tilt",
    "tiltFine",
    "infinitePan",
    "infiniteTilt",
    "movementSpeed",
    "dimmer",
    "dimmerFine",
    "colorWheel",
    "colorComponent",
    "goboWheel",
    "goboIndexing",
    "goboIndexingFine",
    "goboRotation",
    "goboRotationFine",
    "goboShake",
    "shutterStrobe",
    "focus",
    "focusFine",
    "zoom",
    "zoomFine",
    "iris",
    "irisFine",
    "frost",
    "frostFine",
    "prism",
    "prismRotation",
    "fog",
    "custom",
]);

export function readLiveWidgetOverride(props: JSONMap | undefined): DMXLiveWidget | undefined {
    const raw = props?.liveWidget;
    if (typeof raw !== "string") {
        return undefined;
    }
    const v = raw.trim() as DMXLiveWidget;
    return DMX_LIVE_WIDGET_OPTIONS.some((o) => o.value === v) ? v : undefined;
}

function entrySpan(e: { from: number; to: number }): number {
    return Math.abs(e.to - e.from) + 1;
}

function entriesLookDiscrete(entries: ReturnType<typeof parseFixtureEntries>): boolean {
    if (entries.length === 0 || entries.length > 12) {
        return false;
    }
    return entries.every((e) => entrySpan(e) <= 20);
}

function entriesHaveWideRange(entries: ReturnType<typeof parseFixtureEntries>): boolean {
    return entries.some((e) => entrySpan(e) > 20);
}

/** Off slot at 0 plus a volume range — typical smoke/hazer fog channel. */
function entriesLookLikeOffPlusVolume(entries: ReturnType<typeof parseFixtureEntries>): boolean {
    let hasOff = false;
    let hasVolume = false;
    for (const e of entries) {
        const lo = Math.min(e.from, e.to);
        const hi = Math.max(e.from, e.to);
        if (lo === 0 && hi === 0) {
            hasOff = true;
            continue;
        }
        if (hi >= 1) {
            hasVolume = true;
        }
    }
    return hasOff && hasVolume;
}

function hasLinearRange(props: JSONMap | undefined): boolean {
    return typeof props?.min === "number" || typeof props?.max === "number";
}

export function inferLiveWidget(ch: DMXChannel): DMXLiveWidget {
    if (FINE_TYPES.has(ch.type)) {
        return "hidden";
    }
    if (!LIVE_MAPPABLE_TYPES.has(ch.type)) {
        return "hidden";
    }

    const props = ch.properties as JSONMap | undefined;
    const entries = parseFixtureEntries(props);

    switch (ch.type) {
        case "colorWheel":
            return entries.length > 0 ? "colorWheel" : hasLinearRange(props) ? "slider" : "hidden";
        case "goboWheel":
            return entries.length > 0 ? "goboWheel" : "hidden";
        case "shutterStrobe":
            return entries.length > 0 ? "shutterModes" : hasLinearRange(props) ? "slider" : "hidden";
        case "pan":
        case "tilt":
        case "infinitePan":
        case "infiniteTilt":
        case "dimmer":
        case "dimmerFine":
        case "zoom":
        case "focus":
        case "iris":
            if (entries.length > 0) {
                return entriesLookDiscrete(entries) ? "buttons" : "slotSlider";
            }
            return "slider";
        case "fog":
            if (entries.length > 0 && entriesLookLikeOffPlusVolume(entries)) {
                return "buttonSlider";
            }
            if (entries.length > 0) {
                return entriesLookDiscrete(entries) ? "buttons" : "slotSlider";
            }
            return "slider";
        case "frost":
            if (entries.length > 0) {
                return entriesHaveWideRange(entries) ? "slotSlider" : "buttons";
            }
            return "slider";
        case "movementSpeed":
        case "goboRotation":
        case "goboRotationFine":
        case "goboShake":
        case "goboIndexing":
        case "goboIndexingFine":
        case "prism":
        case "prismRotation":
        case "prismIndexing":
        case "prismIndexingFine":
        case "custom":
            if (entries.length > 0) {
                return entriesLookDiscrete(entries) ? "buttons" : "slotSlider";
            }
            return hasLinearRange(props) ? "slider" : "hidden";
        default:
            if (entries.length > 0) {
                return entriesLookDiscrete(entries) ? "buttons" : "slotSlider";
            }
            return hasLinearRange(props) ? "slider" : "hidden";
    }
}

export function resolveLiveWidget(ch: DMXChannel): DMXLiveWidget {
    const override = readLiveWidgetOverride(ch.properties as JSONMap | undefined);
    if (override && override !== "auto") {
        return override;
    }
    return inferLiveWidget(ch);
}

/** Why a channel is omitted from the live tab. */
export type LiveWidgetHiddenSource = "override" | "inferred";

export function isLiveWidgetHidden(ch: DMXChannel): boolean {
    return resolveLiveWidget(ch) === "hidden";
}

/** When hidden: `override` = Live control set to Hidden; `inferred` = Auto (e.g. fine channels). */
export function liveWidgetHiddenSource(ch: DMXChannel): LiveWidgetHiddenSource | null {
    if (!isLiveWidgetHidden(ch)) {
        return null;
    }
    const override = readLiveWidgetOverride(ch.properties as JSONMap | undefined);
    return override === "hidden" ? "override" : "inferred";
}

export function liveWidgetHiddenBadgeLabel(source: LiveWidgetHiddenSource): string {
    return source === "override" ? "Hidden · set in editor" : "Hidden · auto";
}

export function liveWidgetLabel(widget: DMXLiveWidget): string {
    return DMX_LIVE_WIDGET_OPTIONS.find((o) => o.value === widget)?.label ?? widget;
}

/** How the live tab shows the current value for a linear `slider` widget. */
export type LiveSliderLabelMode = "percent" | "dmx";

export const LIVE_SLIDER_LABEL_OPTIONS: { value: LiveSliderLabelMode; label: string }[] = [
    {value: "percent", label: "Percent (%)"},
    {value: "dmx", label: "DMX (0–255)"},
];

export function isDegreeSliderChannel(ch: DMXChannel): boolean {
    return (
        ch.type === "pan" ||
        ch.type === "tilt" ||
        ch.type === "infinitePan" ||
        ch.type === "infiniteTilt"
    );
}

export function readLiveSliderLabelMode(props: JSONMap | undefined, ch: DMXChannel): LiveSliderLabelMode {
    const raw = props?.liveSliderLabel;
    if (raw === "percent" || raw === "dmx") {
        return raw;
    }
    if (ch.type === "dimmer" || ch.type === "dimmerFine") {
        return "percent";
    }
    return "dmx";
}

export function liveSliderLabelModeHint(mode: LiveSliderLabelMode): string {
    return mode === "percent" ? "%" : "0–255";
}

export function channelLiveTileId(ch: DMXChannel): string {
    return `ch-${ch.channel}`;
}

export function parseChannelLiveTileId(id: string): number | null {
    if (!id.startsWith("ch-")) {
        return null;
    }
    const n = Number(id.slice(3));
    return Number.isFinite(n) && n >= 1 ? Math.round(n) : null;
}

function readEntryLiveSlotKind(raw: unknown): LiveSlotKind | undefined {
    if (typeof raw !== "string") {
        return undefined;
    }
    const v = raw.trim();
    return v === "button" || v === "slider" ? v : undefined;
}

function inferEntryLiveSlotKind(
    entry: { from: number; to: number; label?: string; mode?: string },
    index: number,
): LiveSlotKind {
    const lo = Math.min(entry.from, entry.to);
    const hi = Math.max(entry.from, entry.to);
    const span = hi - lo + 1;
    const hay = `${entry.label ?? ""} ${entry.mode ?? ""}`.toLowerCase();
    if (lo === 0 && hi === 0) {
        return "button";
    }
    if (hay.includes("off") && span <= 1) {
        return "button";
    }
    if (span <= 3) {
        return "button";
    }
    return "slider";
}

/** Per-entry live control: switch vs range slider (for `buttonSlider` widget). */
export function parseEntryLiveSlotKinds(
    props: JSONMap | undefined,
    entries: ReturnType<typeof parseFixtureEntries>,
): LiveSlotKind[] {
    const raw = props?.entries;
    const rawList = Array.isArray(raw) ? raw : [];
    return entries.map((entry, i) => {
        const item = rawList[i];
        if (item && typeof item === "object" && !Array.isArray(item)) {
            const fromItem = readEntryLiveSlotKind((item as Record<string, unknown>).liveSlotKind);
            if (fromItem) {
                return fromItem;
            }
        }
        return inferEntryLiveSlotKind(entry, i);
    });
}

export function findOffButtonSlotIndex(
    entries: ReturnType<typeof parseFixtureEntries>,
    kinds: LiveSlotKind[],
): number {
    for (let i = 0; i < entries.length; i++) {
        if (kinds[i] !== "button") {
            continue;
        }
        const e = entries[i];
        const lo = Math.min(e.from, e.to);
        const hi = Math.max(e.from, e.to);
        const hay = `${e.label ?? ""} ${e.mode ?? ""}`.toLowerCase();
        if (lo === 0 && hi === 0) {
            return i;
        }
        if (hay.includes("off")) {
            return i;
        }
    }
    return -1;
}

export function firstSliderSlotIndex(kinds: LiveSlotKind[]): number {
    return kinds.findIndex((k) => k === "slider");
}

export function isWidgetAllowedForChannelType(widget: DMXLiveWidget, type: DMXChannelType): boolean {
    if (
        widget === "auto" ||
        widget === "hidden" ||
        widget === "slider" ||
        widget === "slotSlider" ||
        widget === "buttons" ||
        widget === "buttonSlider"
    ) {
        return true;
    }
    if (widget === "colorWheel") {
        return type === "colorWheel" || type === "colorComponent";
    }
    if (widget === "goboWheel") {
        return type === "goboWheel";
    }
    if (widget === "shutterModes") {
        return type === "shutterStrobe";
    }
    return false;
}

File: internal/controller/dmx_live_widget.go
package controller

import "strings"

type dmxLiveWidget string

const (
	liveWidgetAuto         dmxLiveWidget = "auto"
	liveWidgetHidden       dmxLiveWidget = "hidden"
	liveWidgetSlider       dmxLiveWidget = "slider"
	liveWidgetSlotSlider   dmxLiveWidget = "slotSlider"
	liveWidgetButtons      dmxLiveWidget = "buttons"
	liveWidgetButtonSlider dmxLiveWidget = "buttonSlider"
	liveWidgetColorWheel   dmxLiveWidget = "colorWheel"
	liveWidgetGoboWheel    dmxLiveWidget = "goboWheel"
	liveWidgetShutterModes dmxLiveWidget = "shutterModes"
)

func readLiveWidgetOverride(props map[string]any) dmxLiveWidget {
	if props == nil {
		return liveWidgetAuto
	}
	raw, ok := props["liveWidget"].(string)
	if !ok {
		return liveWidgetAuto
	}
	switch strings.TrimSpace(raw) {
	case "hidden", "slider", "slotSlider", "buttons", "buttonSlider", "colorWheel", "goboWheel", "shutterModes":
		return dmxLiveWidget(strings.TrimSpace(raw))
	default:
		return liveWidgetAuto
	}
}

func isFineChannelType(t string) bool {
	switch strings.ToLower(strings.TrimSpace(t)) {
	case "panfine", "tiltfine", "dimmerfine", "zoomfine", "focusfine", "irisfine", "frostfine", "goborotationfine":
		return true
	default:
		return false
	}
}

func entrySpan(from, to float64) float64 {
	if from <= to {
		return to - from + 1
	}
	return from - to + 1
}

func entriesLookDiscrete(entries []liveFixtureEntry) bool {
	if len(entries) == 0 || len(entries) > 12 {
		return false
	}
	for _, e := range entries {
		if entrySpan(e.from, e.to) > 20 {
			return false
		}
	}
	return true
}

func entriesLookLikeOffPlusVolume(entries []liveFixtureEntry) bool {
	hasOff := false
	hasVolume := false
	for _, e := range entries {
		lo := int(e.from)
		hi := int(e.to)
		if lo > hi {
			lo, hi = hi, lo
		}
		if lo == 0 && hi == 0 {
			hasOff = true
			continue
		}
		if hi >= 1 {
			hasVolume = true
		}
	}
	return hasOff && hasVolume
}

func entriesHaveWideRange(entries []liveFixtureEntry) bool {
	for _, e := range entries {
		if entrySpan(e.from, e.to) > 20 {
			return true
		}
	}
	return false
}

func inferLiveWidget(ch DMXChannel) dmxLiveWidget {
	typ := strings.TrimSpace(ch.Type)
	if isFineChannelType(typ) {
		return liveWidgetHidden
	}
	props := ch.Properties
	entries := liveInitParseEntries(props)

	switch strings.ToLower(typ) {
	case "colorwheel":
		if len(entries) > 0 {
			return liveWidgetColorWheel
		}
		return liveWidgetSlider
	case "gobowheel":
		if len(entries) > 0 {
			return liveWidgetGoboWheel
		}
		return liveWidgetHidden
	case "shutterstrobe":
		if len(entries) > 0 {
			return liveWidgetShutterModes
		}
		return liveWidgetSlider
	case "frost":
		if len(entries) > 0 {
			if entriesHaveWideRange(entries) {
				return liveWidgetSlotSlider
			}
			return liveWidgetButtons
		}
		return liveWidgetSlider
	case "movementspeed", "goborotation", "goborotationfine", "goboshake", "goboindexing", "goboindexingfine",
		"prism", "prismrotation", "prismindexing", "prismindexingfine", "custom":
		if len(entries) > 0 {
			if entriesLookDiscrete(entries) {
				return liveWidgetButtons
			}
			return liveWidgetSlotSlider
		}
		return liveWidgetSlider
	case "fog":
		if len(entries) > 0 && entriesLookLikeOffPlusVolume(entries) {
			return liveWidgetButtonSlider
		}
		if len(entries) > 0 {
			if entriesLookDiscrete(entries) {
				return liveWidgetButtons
			}
			return liveWidgetSlotSlider
		}
		return liveWidgetSlider
	case "pan", "tilt", "infinitepan", "infinitetilt", "dimmer", "zoom", "focus", "iris":
		if len(entries) > 0 {
			if entriesLookDiscrete(entries) {
				return liveWidgetButtons
			}
			return liveWidgetSlotSlider
		}
		return liveWidgetSlider
	default:
		if len(entries) > 0 {
			if entriesLookDiscrete(entries) {
				return liveWidgetButtons
			}
			return liveWidgetSlotSlider
		}
		return liveWidgetHidden
	}
}

func resolveLiveWidget(ch DMXChannel) dmxLiveWidget {
	override := readLiveWidgetOverride(ch.Properties)
	if override != liveWidgetAuto && override != "" {
		return override
	}
	return inferLiveWidget(ch)
}

func findCoarseForFine(fixture DMXFixture, fine *DMXChannel) *DMXChannel {
	if fine == nil {
		return nil
	}
	typ := strings.ToLower(strings.TrimSpace(fine.Type))
	for i := range fixture.Channels {
		c := &fixture.Channels[i]
		switch typ {
		case "panfine":
			if strings.EqualFold(c.Type, "pan") || strings.EqualFold(c.Type, "infinitePan") {
				return c
			}
		case "tiltfine":
			if strings.EqualFold(c.Type, "tilt") || strings.EqualFold(c.Type, "infiniteTilt") {
				return c
			}
		}
	}
	return nil
}

func liveInitFindOffButtonSlot(entries []liveFixtureEntry, kinds []string) int {
	for i := range entries {
		if i >= len(kinds) || kinds[i] != "button" {
			continue
		}
		e := entries[i]
		lo := int(e.from)
		hi := int(e.to)
		if lo > hi {
			lo, hi = hi, lo
		}
		hay := strings.ToLower(e.mode + " " + e.label)
		if lo == 0 && hi == 0 {
			return i
		}
		if strings.Contains(hay, "off") {
			return i
		}
	}
	return -1
}

func liveInitParseEntrySlotKinds(props map[string]any, entries []liveFixtureEntry) []string {
	var raw []any
	if props != nil {
		if r, ok := props["entries"].([]any); ok {
			raw = r
		}
	}
	out := make([]string, len(entries))
	for i := range entries {
		kind := "slider"
		if i < len(raw) {
			if m, ok := raw[i].(map[string]any); ok {
				if s, ok := m["liveSlotKind"].(string); ok {
					s = strings.TrimSpace(s)
					if s == "button" || s == "slider" {
						kind = s
						out[i] = kind
						continue
					}
				}
			}
		}
		e := entries[i]
		lo := int(e.from)
		hi := int(e.to)
		if lo > hi {
			lo, hi = hi, lo
		}
		span := hi - lo + 1
		hay := strings.ToLower(e.mode + " " + e.label)
		if lo == 0 && hi == 0 || strings.Contains(hay, "off") && span <= 1 || span <= 3 {
			kind = "button"
		}
		out[i] = kind
	}
	return out
}

func liveInitFirstSliderSlot(kinds []string) int {
	for i, k := range kinds {
		if k == "slider" {
			return i
		}
	}
	return -1
}

func liveInitButtonSliderByte(props map[string]any, entries []liveFixtureEntry) int {
	if len(entries) == 0 {
		return liveInitLinearByte(props, 0)
	}
	kinds := liveInitParseEntrySlotKinds(props, entries)
	offIdx := liveInitFindOffButtonSlot(entries, kinds)
	if offIdx >= 0 {
		return 0
	}
	sliderIdx := liveInitFirstSliderSlot(kinds)
	if sliderIdx >= 0 {
		return liveInitSlotByte(entries, sliderIdx, 0)
	}
	return 0
}

func liveInitOutputForChannel(fixture DMXFixture, ch *DMXChannel) int {
	if ch == nil {
		return 0
	}
	if v, ok := explicitDefaultOutputByte(ch); ok {
		return v
	}
	widget := resolveLiveWidget(*ch)
	props := ch.Properties
	entries := liveInitParseEntries(props)

	switch widget {
	case liveWidgetHidden:
		coarse := findCoarseForFine(fixture, ch)
		if coarse != nil {
			return liveInitOutputForChannel(fixture, coarse)
		}
		return 0
	case liveWidgetSlider:
		linear := 0.5
		if strings.EqualFold(ch.Type, "dimmer") {
			linear = 1.0
		}
		if strings.EqualFold(ch.Type, "frost") {
			linear = 0.0
		}
		return liveInitLinearByte(props, linear)
	case liveWidgetShutterModes:
		idx := liveInitPickShutterEntryIndex(entries, "open")
		return liveInitSlotMid(entries, idx)
	case liveWidgetColorWheel, liveWidgetGoboWheel, liveWidgetButtons:
		return liveInitSlotMid(entries, 0)
	case liveWidgetSlotSlider:
		if strings.EqualFold(ch.Type, "frost") && len(entries) > 0 {
			pool := liveInitFrostEntriesForCurve(entries, "linear")
			usePool := pool
			if len(usePool) == 0 {
				usePool = entries
			}
			return liveInitSlotMid(usePool, 0)
		}
		if len(entries) > 0 {
			return liveInitSlotByte(entries, 0, 0.5)
		}
		return liveInitLinearByte(props, 0.5)
	case liveWidgetButtonSlider:
		return liveInitButtonSliderByte(props, entries)
	default:
		return liveInitLinearByte(props, 0.5)
	}
}

func explicitDefaultOutputByte(ch *DMXChannel) (int, bool) {
	if ch == nil || ch.DefaultValue == nil {
		return 0, false
	}
	return clampDMXByte(*ch.DefaultValue), true
}
